ambari-server/src/main/resources/Ambari-DDL-Derby-CREATE.sql

```diff
@@ -893,7 +893,7 @@ CREATE TABLE topology_request (
   action VARCHAR(255) NOT NULL,
   cluster_id BIGINT NOT NULL,
   bp_name VARCHAR(100) NOT NULL,
-  raw_request_body CLOB NOT NULL,
+  mpack_instances CLOB NOT NULL,
```

Comment:
Can you give an example of what mpack_instances looks like? Does it make sense to try to normalize this data into its own table?

Reply:
It has the same structure as for blueprints. For blueprints it is modelled by the BlueprintMpackInstanceEntity, BlueprintServiceEntity, BlueprintMpackConfigEntities. 

Currently we are interested in the stack id's only, later also in configurations and service descriptors associated with mpacks (I saw configs are dumped as json too).

Do you think it is worth normalizing? (would mean 3 more tables, later maybe more)

Reply:
My concern is always that when de-serializing this, if the object it's being mapped to changed in any way, the deserialization will fail? We've actually tried to move away from this model. Repositories used to be a solid chunk of JSON but we recently broke this out into 3 new tables to better manage it as entities. 

I hate giant blob of JSON, especially if we need to worry about deserializing it ourselves and pulling data out of it.

Reply:
Based on @jonathan-hurley 's feedback, and also based on looking at the BlueprintMPackInstanceEntity, it seems like this data should be normalized.  

Since the MPack entities can be defined in either the Blueprint or the cluster creation template, it seems like it might make sense to use the same tables to store them.  

Is there some technical reason for storing this data in the topology_request table?  It seems to me that we'd be better off persisting all MPack instance data in the same table, regardless of whether it is defined in the Blueprint or the Cluster Creation template. 

Since Service instance definitions and config can be specified in either document, I'm not sure why these should be modelled separately based on the location of the definition.  It looks like we should be using the existing tables for both cases.

Reply:
On the other hand, this shouldn't be a problem for blueprints, since backward compatibility is required.  If users should be able to successfully submit "old" blueprint / cluster creation requests even if code is changed, so should Ambari internally.